src/dependency-manager/spec/database-spec.ts

```diff
@@ -0,0 +1,37 @@
+import { IsOptional, IsString } from 'class-validator';
+import { JSONSchema } from 'class-validator-jsonschema';
+import { AnyOf, ExpressionOr } from './utils/json-schema-annotations';
+import { Slugs } from './utils/slugs';
+
+export class DatabaseSpec {
+  @IsOptional()
+  @JSONSchema({
+    type: 'string',
+    pattern: Slugs.ArchitectSlugValidator.source,
+    errorMessage: Slugs.ArchitectSlugDescription,
+    description: 'A specific service name which will override the database name specified in the component.',
+  })
+  reserved_name?: string;
+
+  @IsOptional()
+  @JSONSchema({
+    type: 'string',
+    description: 'Human readable description',
+  })
+  description?: string;
+
+  @IsString()
+  @JSONSchema({
+    ...ExpressionOr({ type: 'string', pattern: Slugs.ComponentDatabaseValidator.source }),
+    description: 'The type engine and version of database software needed for data storage.',
+    errorMessage: Slugs.ComponentDatabaseDescription,
+  })
+  type!: string;
+
+  @IsOptional()
+  @JSONSchema({
+    ...ExpressionOr(AnyOf('null', 'string')),
+    description: 'The connection string of an existing database to use instead of provisioning a new one',
+  })
+  dsn?: null | string;
```

Comment:
On this note - I think `url` might be better here instead of `dsn`

Reply:
I think most of the time DSN is referenced as this format:

```
<protocol>://<user>:<pass>@<host>:<port>/<database>
```

while URLs are usually expected to be this:

```
<protocol>://<host>:<port>
```

I know that's not exactly what `url` means for service interfaces today since we also optionally inject the username and password, but I thought the added path made it feel different. That being said, we could definitely have both terms available for both resource types (services and DBs) and just favor different language in our docs when talking about them based on what developers are expecting or searching for naturally.

Reply:
Thinking about this more I think we should change it to `connectionString`. DSN is generally used in places where you have the configuration stored somewhere else and are just referencing the configuration by name. While it can be used as the connection string, in my experience dsn is not as well understood anymore.

The other thing is that there are generally two types of connection strings. The first is your standard url but then other systems like .NET actually use ODBC which uses its own connection string syntax. To make sure we support as many formats as possible we should also add other values that can be used in interpolation.

So a user may configure a `connectionString` as part of their override, but they should be able to access it for ODBC connections via `connectionStringODBC`. The same thing should be allowed in the reverse. Otherwise, you may need to declare two databases each with its own connection string in order to support every use case.

Reply:
`connectionString` is a bit wordy, but its the only other keyword that I've thought makes sense. We can make that the input value if you want.

For outputs (e.g. referencing `${{ databases.name.connectionString }}`) I think we should support both names since that allows someone to be brief or verbose. Should be easy enough to implement too.